### ISSUE TITLE
"require-returns" requires useless return statements for return type "undefined"

### ISSUE THREAD

**Issue by lll000111** [gajus/eslint-plugin-jsdoc#142]
Subject says it. If the return type is undefined I don't need a return statement in the code.

This plugin should concern itself with JSDoc &mdash; not with the code style! The return type when there is no `return` statement is `undefined`, so describing such a function as having `@returns {undefined}` is 100% correct. I should not get

    Present JSDoc @returns declaration but not available return expression in function  jsdoc/require-returns

**Comment by gajus** [comment]
PR is welcome.

**Comment by lll000111** [comment]
Sure, always. And so are big reports, and here is one.

**Comment by ibbignerd** [comment]
[According to ESLint](https://eslint.org/docs/rules/valid-jsdoc), `@returns {void}` should be valid.

[According to Google Closure Compiler](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#return-type-description), omitting it completely or `@returns {undefined]` should be valid. The following should all be tests that satisfy these requirements.

```javascript
valid: [
    {
      code: `
          /**
           * @returns {Object}
           */
          function quux () {

            return {a: foo};
          }
      `
    },
    {
      code: `
          /**
           * @returns {Object}
           */
          const quux = () => ({a: foo});
      `
    },
    {
      code: `
          /**
           * @returns {Object}
           */
          const quux = () => {
            return {a: foo}
          };
      `
    },
    {
      code: `
          /**
           * @returns {void}
           */
          function quux () {
          }
      `
    },
    {
      code: `
          /**
           * @returns {void}
           */
          const quux = () => {
            
          }
      `
    },
    {
      code: `
          /**
           * @returns {undefined}
           */
          function quux () {
          }
      `
    },
    {
      code: `
          /**
           * @returns {undefined}
           */
          const quux = () => {
            
          }
      `
    },
    {
      code: `
          /**
           *
           */
          function quux () {
          }
      `
    },
    {
      code: `
          /**
           *
           */
          const quux = () => {
            
          }
      `
    }
  ]
```

**Comment by ChrisHSandN** [comment]
It's no PR, but might be helpful. This is the change I made to work around this bug and get my code validating:

https://github.com/gajus/eslint-plugin-jsdoc/blob/master/src/rules/requireReturns.js

```javascript
// replace
if (JSON.stringify(jsdocTags) !== '[]' && sourcecode.indexOf('return') < 1) {
// with
if (JSON.stringify(jsdocTags) !== '[]' && jsdocTags.findIndex(v => ['undefined', 'void'].indexOf(v.type) !== -1) === -1 && sourcecode.indexOf('return') < 1) {
```

Lots of the core ESLint rules seem to have a configurable options object to enable/disable parts of a rule, As this rule has 2 report() cases maybe it would be logical to allow them to be enabled separately?

**Comment by gajus** [comment]
> Lots of the core ESLint rules seem to have a configurable options object to enable/disable parts of a rule, As this rule has 2 report() cases maybe it would be logical to allow them to be enabled separately?

Makes sense.

**Comment by ptkdev** [comment]
Hi,
Sorry for this issue. I fix this in a pull request. 
I use @ChrisHSandN path (thanks, imho is ok) and split this rule in two rules.

**Comment by brettz9** [comment]
FWIW, I'm glad the old rule was kept as is (while adding a new rule to address the concern raised in this issue), because requiring `@returns` in all cases can have some practical utility in flagging to a development team or to users that the function's return has indeed been considered and that it is not merely due to a failure to document.

It seems like this issue is now resolved?

**Comment by brettz9** [comment]
My apologies--I misunderstood the additional rule--I would hope that the old behavior could be restored and the new behavior made optional because I personally *always* want returns documented so as to flag there has been no omission in consideration of return values for documentation.
